File: src/test/java/com/browserselector/util/PatternMatcherTest.java
package com.browserselector.util;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.assertj.core.api.Assertions.assertThat;

class PatternMatcherTest {

    @Nested
    @DisplayName("matches() domain patterns")
    class MatchesDomainPatterns {

        @ParameterizedTest
        @CsvSource({
            "google.com, https://google.com, true",
            "google.com, http://google.com, true",
            "google.com, https://google.com/search, true",
            "example.org, https://example.org/page, true"
        })
        @DisplayName("exact domain pattern matches domain")
        void exactDomainMatches(String pattern, String url, boolean expected) {
            assertThat(PatternMatcher.matches(pattern, url)).isEqualTo(expected);
        }

        @ParameterizedTest
        @CsvSource({
            "google.com, https://www.google.com, true",
            "google.com, https://mail.google.com, true",
            "google.com, https://docs.google.com/document, true",
            "example.org, https://blog.example.org, true"
        })
        @DisplayName("domain pattern matches subdomains")
        void domainMatchesSubdomains(String pattern, String url, boolean expected) {
            assertThat(PatternMatcher.matches(pattern, url)).isEqualTo(expected);
        }

        @ParameterizedTest
        @CsvSource({
            "*.google.com, https://mail.google.com, true",
            "*.google.com, https://docs.google.com, true",
            "*.google.com, https://www.google.com, true",
            "*.example.org, https://api.example.org/v1, true"
        })
        @DisplayName("wildcard pattern matches subdomains")
        void wildcardMatchesSubdomains(String pattern, String url, boolean expected) {
            assertThat(PatternMatcher.matches(pattern, url)).isEqualTo(expected);
        }

        @ParameterizedTest
        @CsvSource({
            "*.google.com, https://google.com, true",
            "*.example.org, https://example.org, true",
            "*.github.com, https://github.com/repo, true"
        })
        @DisplayName("wildcard pattern also matches bare domain")
        void wildcardMatchesBareDomain(String pattern, String url, boolean expected) {
            assertThat(PatternMatcher.matches(pattern, url)).isEqualTo(expected);
        }

        @ParameterizedTest
        @CsvSource({
            "google.com, https://bing.com, false",
            "google.com, https://notgoogle.com, false",
            "google.com, https://google.org, false",
            "*.google.com, https://google.org, false",
            "example.org, https://example.com, false"
        })
        @DisplayName("pattern does not match different domains")
        void noMatchDifferentDomain(String pattern, String url, boolean expected) {
            assertThat(PatternMatcher.matches(pattern, url)).isEqualTo(expected);
        }
    }

    @Nested
    @DisplayName("matches() path patterns")
    class MatchesPathPatterns {

        @Test
        @DisplayName("double wildcard matches path segments")
        void doubleWildcardMatchesPath() {
            assertThat(PatternMatcher.matches("example.com/**", "https://example.com/path/to/file")).isTrue();
            assertThat(PatternMatcher.matches("example.com/api/**", "https://example.com/api/v1/users")).isTrue();
        }

        @Test
        @DisplayName("single wildcard in path matches segment")
        void singleWildcardInPath() {
            assertThat(PatternMatcher.matches("example.com/*.html", "https://example.com/page.html")).isTrue();
        }

        @Test
        @DisplayName("question mark matches single character")
        void questionMarkMatchesSingleChar() {
            assertThat(PatternMatcher.matches("example.co?", "https://example.com")).isTrue();
            assertThat(PatternMatcher.matches("exampl?.com", "https://example.com")).isTrue();
        }
    }

    @Nested
    @DisplayName("matches() edge cases")
    class MatchesEdgeCases {

        @Test
        @DisplayName("null inputs return false")
        void nullInputs() {
            assertThat(PatternMatcher.matches(null, "https://google.com")).isFalse();
            assertThat(PatternMatcher.matches("google.com", null)).isFalse();
            assertThat(PatternMatcher.matches(null, null)).isFalse();
        }
    }

    @Nested
    @DisplayName("isValidPattern()")
    class IsValidPattern {

        @ParameterizedTest
        @CsvSource({
            "google.com, true",
            "*.google.com, true",
            "**.google.com, true",
            "example.org/path/*, true",
            "https://example.com, true",
            "http://example.com, true",
            "?.google.com, true"
        })
        @DisplayName("returns true for valid patterns")
        void validPatterns(String pattern, boolean expected) {
            assertThat(PatternMatcher.isValidPattern(pattern)).isEqualTo(expected);
        }

        @Test
        @DisplayName("returns false for blank/null patterns")
        void invalidPatterns() {
            assertThat(PatternMatcher.isValidPattern(null)).isFalse();
            assertThat(PatternMatcher.isValidPattern("")).isFalse();
            assertThat(PatternMatcher.isValidPattern("   ")).isFalse();
        }

        @Test
        @DisplayName("returns false for wildcard-only patterns")
        void wildcardOnlyPatterns() {
            assertThat(PatternMatcher.isValidPattern("***")).isFalse();
            assertThat(PatternMatcher.isValidPattern("???")).isFalse();
            assertThat(PatternMatcher.isValidPattern("*?*")).isFalse();
        }
    }

    @Nested
    @DisplayName("domainToPattern()")
    class DomainToPattern {

        @Test
        @DisplayName("returns domain as-is (no conversion)")
        void returnsDomainAsIs() {
            assertThat(PatternMatcher.domainToPattern("google.com")).isEqualTo("google.com");
            assertThat(PatternMatcher.domainToPattern("mail.google.com")).isEqualTo("mail.google.com");
        }

        @Test
        @DisplayName("preserves existing wildcards")
        void preservesWildcards() {
            assertThat(PatternMatcher.domainToPattern("*.google.com")).isEqualTo("*.google.com");
            assertThat(PatternMatcher.domainToPattern("example.com/**")).isEqualTo("example.com/**");
        }

        @Test
        @DisplayName("handles blank input")
        void blankInput() {
            assertThat(PatternMatcher.domainToPattern(null)).isEqualTo("");
            assertThat(PatternMatcher.domainToPattern("")).isEqualTo("");
            assertThat(PatternMatcher.domainToPattern("   ")).isEqualTo("");
        }
    }
}

File: android/app/src/test/java/com/browserselector/android/util/PatternMatcherTest.kt
package com.browserselector.android.util

import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for PatternMatcher utility.
 */
class PatternMatcherTest {

    // ==================== Domain Pattern Tests ====================

    @Test
    fun `domain pattern matches exact domain`() {
        assertTrue(PatternMatcher.matches("google.com", "https://google.com"))
        assertTrue(PatternMatcher.matches("google.com", "http://google.com"))
        assertTrue(PatternMatcher.matches("google.com", "https://google.com/search"))
        assertTrue(PatternMatcher.matches("example.org", "https://example.org/page"))
    }

    @Test
    fun `domain pattern matches subdomain`() {
        assertTrue(PatternMatcher.matches("google.com", "https://www.google.com"))
        assertTrue(PatternMatcher.matches("google.com", "https://mail.google.com"))
        assertTrue(PatternMatcher.matches("google.com", "https://docs.google.com/document"))
        assertTrue(PatternMatcher.matches("example.org", "https://blog.example.org"))
    }

    @Test
    fun `wildcard pattern matches subdomain`() {
        assertTrue(PatternMatcher.matches("*.google.com", "https://mail.google.com"))
        assertTrue(PatternMatcher.matches("*.google.com", "https://docs.google.com"))
        assertTrue(PatternMatcher.matches("*.google.com", "https://www.google.com"))
        assertTrue(PatternMatcher.matches("*.example.org", "https://api.example.org/v1"))
    }

    @Test
    fun `wildcard pattern matches bare domain`() {
        // *.domain.com should also match domain.com (without subdomain)
        assertTrue(PatternMatcher.matches("*.google.com", "https://google.com"))
        assertTrue(PatternMatcher.matches("*.example.org", "https://example.org"))
        assertTrue(PatternMatcher.matches("*.github.com", "https://github.com/repo"))
    }

    @Test
    fun `pattern does not match different domain`() {
        assertFalse(PatternMatcher.matches("google.com", "https://bing.com"))
        assertFalse(PatternMatcher.matches("google.com", "https://notgoogle.com"))
        assertFalse(PatternMatcher.matches("google.com", "https://google.org"))
        assertFalse(PatternMatcher.matches("*.google.com", "https://google.org"))
        assertFalse(PatternMatcher.matches("example.org", "https://example.com"))
    }

    @Test
    fun `blank inputs return false`() {
        assertFalse(PatternMatcher.matches("", "https://google.com"))
        assertFalse(PatternMatcher.matches("google.com", ""))
        assertFalse(PatternMatcher.matches("", ""))
        assertFalse(PatternMatcher.matches("   ", "https://google.com"))
        assertFalse(PatternMatcher.matches("google.com", "   "))
        assertFalse(PatternMatcher.matches("  ", "  "))
    }

    // ==================== extractDomain Tests ====================

    @Test
    fun `extractDomain returns correct domain`() {
        assertEquals("google.com", PatternMatcher.extractDomain("https://google.com"))
        assertEquals("google.com", PatternMatcher.extractDomain("https://www.google.com"))
        assertEquals("google.com", PatternMatcher.extractDomain("http://google.com/search?q=test"))
        assertEquals("mail.google.com", PatternMatcher.extractDomain("https://mail.google.com"))
        assertEquals("example.org", PatternMatcher.extractDomain("https://example.org/page/sub"))
    }

    @Test
    fun `extractDomain handles urls without scheme`() {
        assertEquals("google.com", PatternMatcher.extractDomain("google.com"))
        assertEquals("google.com", PatternMatcher.extractDomain("www.google.com"))
    }

    // ==================== domainToPattern Tests ====================

    @Test
    fun `domainToPattern removes www prefix`() {
        assertEquals("google.com", PatternMatcher.domainToPattern("www.google.com"))
        assertEquals("example.org", PatternMatcher.domainToPattern("www.example.org"))
        assertEquals("github.com", PatternMatcher.domainToPattern("www.github.com"))
    }

    @Test
    fun `domainToPattern preserves domain without www`() {
        assertEquals("google.com", PatternMatcher.domainToPattern("google.com"))
        assertEquals("mail.google.com", PatternMatcher.domainToPattern("mail.google.com"))
    }

    @Test
    fun `domainToPattern trims whitespace`() {
        assertEquals("google.com", PatternMatcher.domainToPattern("  google.com  "))
        assertEquals("google.com", PatternMatcher.domainToPattern("  www.google.com  "))
    }

    // ==================== isValidPattern Tests ====================

    @Test
    fun `isValidPattern returns true for valid patterns`() {
        assertTrue(PatternMatcher.isValidPattern("google.com"))
        assertTrue(PatternMatcher.isValidPattern("*.google.com"))
        assertTrue(PatternMatcher.isValidPattern("**.google.com"))
        assertTrue(PatternMatcher.isValidPattern("example.org/path/*"))
        assertTrue(PatternMatcher.isValidPattern("https://example.com"))
        assertTrue(PatternMatcher.isValidPattern("http://example.com"))
        assertTrue(PatternMatcher.isValidPattern("?.google.com"))
    }

    @Test
    fun `isValidPattern returns false for blank patterns`() {
        assertFalse(PatternMatcher.isValidPattern(""))
        assertFalse(PatternMatcher.isValidPattern("   "))
    }

    // ==================== Additional Wildcard Tests ====================

    @Test
    fun `double wildcard matches path segments`() {
        assertTrue(PatternMatcher.matches("example.com/**", "https://example.com/path/to/file"))
        assertTrue(PatternMatcher.matches("example.com/api/**", "https://example.com/api/v1/users"))
    }

    @Test
    fun `single wildcard does not match path separator`() {
        // Single * should match within path segment only
        assertTrue(PatternMatcher.matches("example.com/*.html", "https://example.com/page.html"))
    }

    @Test
    fun `question mark matches single character`() {
        assertTrue(PatternMatcher.matches("example.co?", "https://example.com"))
        assertTrue(PatternMatcher.matches("exampl?.com", "https://example.com"))
    }

    // ==================== URL Normalization Tests ====================

    @Test
    fun `normalizeUrl adds https scheme when missing`() {
        assertEquals("https://google.com", PatternMatcher.normalizeUrl("google.com"))
        assertEquals("https://example.org", PatternMatcher.normalizeUrl("example.org"))
    }

    @Test
    fun `normalizeUrl preserves existing scheme`() {
        assertEquals("https://google.com", PatternMatcher.normalizeUrl("https://google.com"))
        assertEquals("http://google.com", PatternMatcher.normalizeUrl("http://google.com"))
    }

    @Test
    fun `normalizeUrl trims whitespace`() {
        assertEquals("https://google.com", PatternMatcher.normalizeUrl("  google.com  "))
    }

    // ==================== isValidUrl Tests ====================

    @Test
    fun `isValidUrl returns true for valid urls`() {
        assertTrue(PatternMatcher.isValidUrl("https://google.com"))
        assertTrue(PatternMatcher.isValidUrl("http://example.org"))
        assertTrue(PatternMatcher.isValidUrl("google.com"))
    }

    @Test
    fun `isValidUrl returns false for invalid urls`() {
        assertFalse(PatternMatcher.isValidUrl(""))
        assertFalse(PatternMatcher.isValidUrl("   "))
    }

    // ==================== truncateUrl Tests ====================

    @Test
    fun `truncateUrl returns full url when under limit`() {
        val shortUrl = "https://example.com"
        assertEquals(shortUrl, PatternMatcher.truncateUrl(shortUrl, 60))
    }

    @Test
    fun `truncateUrl truncates long urls with ellipsis`() {
        val longUrl = "https://example.com/very/long/path/to/some/deeply/nested/resource/file.html"
        val result = PatternMatcher.truncateUrl(longUrl, 30)
        assertEquals(30, result.length)
        assertTrue(result.endsWith("..."))
    }
}
